src/matrix.js

```diff
@@ -24,53 +24,150 @@ SVG.Matrix = SVG.invent({
 
   // Add methods
   extend: {
-    // Extract individual transformations
-    extract: function () {
-      // find delta transform points
-      var px = deltaTransformPoint(this, 0, 1)
-      var py = deltaTransformPoint(this, 1, 0)
-      var skewX = 180 / Math.PI * Math.atan2(px.y, px.x) - 90
 
+    // Clones this matrix
+    clone: function () {
+      return new SVG.Matrix(this)
+    },
+
+    // Transform a matrix into another matrix by manipulating the space
+    transform: function (o) {
+      // Get all of the parameters required to form the matrix
+      var flipX = o.flip && (o.flip === 'x' || o.flip === 'both') ? -1 : 1
+      var flipY = o.flip && (o.flip === 'y' || o.flip === 'both') ? -1 : 1
+      var skewX = o.skew && o.skew.length ? o.skew[0]
+        : isFinite(o.skew) ? o.skew
+        : isFinite(o.skewX) ? o.skewX
+        : 0
+      var skewY = o.skew && o.skew.length ? o.skew[1]
+        : isFinite(o.skew) ? o.skew
+        : isFinite(o.skewY) ? o.skewY
+        : 0
+      var scaleX = o.scale && o.scale.length ? o.scale[0] * flipX
+        : isFinite(o.scale) ? o.scale * flipX
+        : isFinite(o.scaleX) ? o.scaleX * flipX
+        : flipX
+      var scaleY = o.scale && o.scale.length ? o.scale[1] * flipY
+        : isFinite(o.scale) ? o.scale * flipY
+        : isFinite(o.scaleY) ? o.scaleY * flipY
+        : flipY
+      var shear = o.shear || 0
+      var theta = o.rotate || 0
+      var ox = o.origin && o.origin.length ? o.origin[0] : o.ox || 0
+      var oy = o.origin && o.origin.length ? o.origin[1] : o.oy || 0
+      var px = o.position && o.position.length ? o.position[0] : o.px
+      var py = o.position && o.position.length ? o.position[1] : o.py
+      var tx = o.translate && o.translate.length ? o.translate[0] : o.tx || 0
+      var ty = o.translate && o.translate.length ? o.translate[1] : o.ty || 0
+      var rx = o.relative && o.relative.length ? o.relative[0] : o.rx || 0
+      var ry = o.relative && o.relative.length ? o.relative[1] : o.ry || 0
+      var currentTransform = new SVG.Matrix(this)
+
```

Comment:
what happens when passing an actual point to origin, position and so on?
I would suggest using a point to get what you need from the input. Something along these lines:

```js
var p = new SVG.Point(o.ox == null ? origin : o.ox, o.oy)
var ox = p.x
var oy = o.y
```

That syntax should allow for:

- `origin: [10, 20]`
- `origin: {x: 10, y: 20}`
- `ox: 10, ox: 20`
- `undefined`

Reply:
Done